File: zero_latency/sysclock.go
package main

import (
	"context"
	"log"

	"github.com/andrew-solarstorm/yellowstone-grpc-client-go"
	pb "github.com/andrew-solarstorm/yellowstone-grpc-client-go/proto"
	bin "github.com/gagliardetto/binary"
	"github.com/gagliardetto/solana-go"
	lru "github.com/hashicorp/golang-lru/v2"
)

// ref: https://docs.rs/solana-program/latest/solana_program/sysvar/clock/struct.Clock.html
type SysVarClock struct {
	Slot                uint64
	EpochStartTimestamp int64
	Epoch               uint64
	LeaderScheduleEpoch uint64
	UnixTimestamp       int64
}

func decodeClock(data []byte) (*SysVarClock, error) {
	var clock SysVarClock

	decoder := bin.NewBinDecoder(data)

	err := decoder.Decode(&clock)
	if err != nil {
		return nil, err
	}

	return &clock, nil
}

type SystemClock struct {
	cli   *yellowstone.GeyserGrpcClient
	slots *lru.Cache[uint64, int64]
}

func NewSysClock() *SystemClock {
	cache, _ := lru.New[uint64, int64](30)
	return &SystemClock{
		slots: cache,
	}
}

func (svc *SystemClock) TimeStamp(slot uint64) int64 {
	slotTime, _ := svc.slots.Get(slot)
	return slotTime
}

func (svc *SystemClock) Close() {
	if svc.cli != nil {
		svc.cli.Close()
	}
}

func (svc *SystemClock) subscribe(endpoint, token string, commitment *pb.CommitmentLevel) {
	builder, err := yellowstone.BuildFromShared(endpoint)
	if err != nil {
		log.Fatalf("Error building client: %v", err)
	}

	clientBuilder := builder.XToken(token).KeepAliveWhileIdle(true)

	grpcClient, err := clientBuilder.Connect(context.Background())
	if err != nil {
		log.Fatalf("Error connecting: %v", err)
	}

	svc.cli = grpcClient

	req := &pb.SubscribeRequest{
		Accounts: map[string]*pb.SubscribeRequestFilterAccounts{
			"sysvar_filter": {
				Account: []string{
					solana.SysVarClockPubkey.String(),
				},
			},
		},
	}

	ctx := context.Background()
	stream, err := grpcClient.SubscribeWithRequest(ctx, req)
	if err != nil {
		log.Fatalf("Error subscribing to geyser: %v", err)
	}

	go grpcClient.Start(stream, func(update *pb.SubscribeUpdate) error {
		switch update.GetUpdateOneof().(type) {
		case *pb.SubscribeUpdate_Account:
			accountUpdate := update.GetAccount()
			account := accountUpdate.Account

			clock, err := decodeClock(account.Data)
			if err != nil {
				return nil
			}

			if _, ok := svc.slots.Get(clock.Slot); !ok {
				svc.slots.Add(clock.Slot, clock.UnixTimestamp)
			}

		default:
			return nil
		}
		return nil
	})
}

File: tx_sysvar_js/sysclock.js
const yellowstone = require('@triton-one/yellowstone-grpc');

function decodeClock(data) {
    if (!data || data.length < 40) return null;
    const view = new DataView(
        data.buffer || data,
        data.byteOffset || 0,
        data.byteLength || data.length
    );
    const slot = Number(view.getBigUint64(0, true));
    const epochStartTimestamp = Number(view.getBigInt64(8, true));
    const epoch = Number(view.getBigUint64(16, true));
    const leaderScheduleEpoch = Number(view.getBigUint64(24, true));
    const unixTimestamp = Number(view.getBigInt64(32, true));
    return { slot, epochStartTimestamp, epoch, leaderScheduleEpoch, unixTimestamp };
}

class LRUCache {
    constructor(capacity) {
        this.capacity = capacity;
        this.cache = new Map();
    }
    get(key) {
        if (!this.cache.has(key)) return 0; // Go lru returns 0 value default for absent ints
        const val = this.cache.get(key);
        this.cache.delete(key);
        this.cache.set(key, val);
        return val;
    }
    has(key) {
        return this.cache.has(key);
    }
    add(key, value) {
        if (this.cache.has(key)) {
            this.cache.delete(key);
        } else if (this.cache.size >= this.capacity) {
            const firstKey = this.cache.keys().next().value;
            this.cache.delete(firstKey);
        }
        this.cache.set(key, value);
    }
}

class SystemClock {
    constructor() {
        this.cli = null;
        this.slots = new LRUCache(30);
    }

    TimeStamp(slot) {
        return this.slots.get(slot) || 0;
    }

    Close() {
        if (this.cli) {
            this.cli.close && this.cli.close();
            this.cli = null;
        }
    }

    async subscribe(endpoint, token, commitment) {
        let commitmentLevel;
        if (commitment === 'FINALIZED') {
            commitmentLevel = yellowstone.CommitmentLevel.FINALIZED;
        } else if (commitment === 'CONFIRMED') {
            commitmentLevel = yellowstone.CommitmentLevel.CONFIRMED;
        } else {
            commitmentLevel = yellowstone.CommitmentLevel.PROCESSED;
        }

        try {
            const client = new yellowstone.default(endpoint, token, {
                "grpc.max_receive_message_length": 64 * 1024 * 1024,
                "grpc.keepalive_time_ms": 10000,
            });
            this.cli = client;
            await this.cli.connect();

            const stream = await client.subscribe();

            const req = {
                slots: {},
                accounts: {
                    sysvar_filter: {
                        account: ["SysvarC1ock11111111111111111111111111111111"],
                        owner: [],
                        filters: []
                    }
                },
                transactions: {},
                transactionsStatus: {},
                entry: {},
                blocksMeta: {},
                accountsDataSlice: [],
                ping: undefined,
                blocks: {},
                commitment: commitmentLevel
            };

            await new Promise((resolve, reject) => {
                stream.write(req, (err) => {
                    if (err) {
                        reject(err);
                    } else {
                        resolve();
                    }
                });
            });

            stream.on('data', (update) => {
                if (!update || !update.account || !update.account.account) return;
                
                const clock = decodeClock(update.account.account.data);
                if (clock) {
                    if (!this.slots.has(clock.slot)) {
                        this.slots.add(clock.slot, clock.unixTimestamp);
                    }
                }
            });

            stream.on('error', (err) => {
                console.error("SystemClock stream error:", err);
            });

        } catch (err) {
            console.error("Error subscribing system clock:", err);
            process.exit(1);
        }
    }
}

module.exports = { SystemClock };
